scanpointgenerator/generators/arraygenerator.py

```diff
@@ -0,0 +1,78 @@
+from scanpointgenerator.compat import np
+from scanpointgenerator.core import Generator
+
+
+def to_list(value):
+    if isinstance(value, list):
+        return value
+    else:
+        return [value]
+
+@Generator.register_subclass("scanpointgenerator:generator/ArrayGenerator:1.0")
+class ArrayGenerator(Generator):
+    """Generate points fron a given list of positions"""
+
+    def __init__(self, axis, units, points, alternate=False):
+        """
+        Args:
+            axis (str)): The scannable axis name
+            units (str): The scannable units.
+            points (list(double)): array positions
+            alternate (bool): Alternate directions
+        """
+
+        self.alternate = alternate
+        self.points = np.array(points, dtype=np.float64)
+        self.size = len(self.points)
+        self.axes = to_list(axis) # to_list just in case an array is passed
+        self.units = {d:u for (d,u) in zip(self.axes, to_list(units))}
+        if len(self.axes) != 1:
+            raise ValueError("ArrayGenerator only supports a single scannable")
+
+    def prepare_arrays(self, index_array):
+        points = self.points
+        # add linear extension to ends of points, representing t=-1 and t=N+1
+        v_left = points[0] - (points[1] - points[0])
+        v_right = points[-1] + (points[-1] - points[-2])
+        shape = points.shape
+        shape = (shape[0] + 2,) + shape[1:]
+        extended = np.empty(shape, dtype=points.dtype)
+        extended[1:-1] = points
+        extended[0] = v_left
+        extended[-1] = v_right
+        points = extended
+        index_floor = np.floor(index_array).astype(np.int32)
+        epsilon = index_array - index_floor
+        index_floor += 1
+        values = points[index_floor] + epsilon * (points[index_floor+1] - points[index_floor])
+        return {self.axes[0]:values}
+
+    def to_dict(self):
+        """Serialize ArrayGenerator to dictionary"""
+
+        d = {
+                "typeid":self.typeid,
+                "axes":self.axes,
+                "units":[self.units[a] for a in self.axes],
+                "points":self.points.tolist(),
+                "alternate":self.alternate,
+            }
+        return d
+
+    @classmethod
+    def from_dict(cls, d):
+        """
+        Create a ArrayGenerator from serialized form.
+
+        Args:
+            d (dict): Serialized generator
+
+        Returns:
+            ArrayGenerator: New ArrayGenerator instance
+        """
+
+        axes = d["axes"]
```

Comment:
This requires quite a few changes in CompoundGenerator too, which looks at generator.axes quite frequently.